<a href="https://colab.research.google.com/github/AshTayade/Kaggle_Higgs_bosson/blob/main/sem2_project_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Notebook 3

##After trying a bagging ensemble technique we are going to try a boosting ensemble technique.We chose Adaboost because we wanted to experiment with different learning rates. Adaboost is not prone to overfitting and we slightly overfit random forest to get better accuracy hence Adaboost would be a good model to compare with Random forest model.

#Adaboost

In [ ]:
#importing libraries
%matplotlib inline
import numpy as np
import matplotlib
import matplotlib.pyplot as plt
import matplotlib.mlab as mlab
from matplotlib import cm
import pandas as pd
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.pipeline import Pipeline
from sklearn.model_selection import GridSearchCV
from sklearn.ensemble import BaggingClassifier, AdaBoostClassifier
from matplotlib import colors
from sklearn.model_selection import cross_val_score 
from sklearn.metrics import confusion_matrix 
from sklearn.metrics import plot_confusion_matrix
from sklearn.linear_model import LogisticRegression
from sklearn import tree
from sklearn.inspection import permutation_importance
from sklearn.model_selection import RandomizedSearchCV
from sklearn.metrics import accuracy_score
from sklearn.metrics import f1_score, make_scorer
from sklearn import metrics,preprocessing, tree
from sklearn.ensemble import AdaBoostClassifier
import pickle
import time


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


Getting our pickled data from drive

In [ ]:
filename = '/content/drive/MyDrive/Higgs_boson/list_dataset.pickle'
dataset_list = pickle.load(open(filename, 'rb'))
X_train =dataset_list[0]
y_train  =dataset_list[1]
X_test = dataset_list[2]
y_test =dataset_list[3]

In [ ]:
#X_test.shape

(75000, 23)

In [ ]:
boost = AdaBoostClassifier() # adaboost instance

In [ ]:
param_grid = {'n_estimators':[100,200,300], 'learning_rate':[0.01, 0.1, 1],} #param grid dictionary

In [ ]:
gridcv = GridSearchCV(boost, param_grid,cv=3,n_jobs=-1 ) #gridsearchcv object

####Running gridcv fit code on another powerful machine and pickling the grid results then importing them into this google colab Notebook

In [ ]:
start_time = time.time()
gridcv.fit(X_train,y_train)
stop_time = time.time()
print("Code started at : ",time.ctime(start_time))
print("Code started at : ",time.ctime(stop_time))
print("Total time taken : ",((stop_time-start_time)/60),"minutes")

Above cell ran in another machine for computation speed

In [ ]:
filename = '/content/drive/MyDrive/Higgs_boson/ada_grid.pickle'
best_ada = pickle.load(open(filename, 'rb'))

/usr/local/lib/python3.7/dist-packages/sklearn/base.py:318: UserWarning: Trying to unpickle estimator DecisionTreeClassifier from version 0.23.2 when using version 0.22.2.post1. This might lead to breaking code or invalid results. Use at your own risk.
  UserWarning)
/usr/local/lib/python3.7/dist-packages/sklearn/base.py:318: UserWarning: Trying to unpickle estimator AdaBoostClassifier from version 0.23.2 when using version 0.22.2.post1. This might lead to breaking code or invalid results. Use at your own risk.
  UserWarning)


Following is our best adaboost model 

In [ ]:
best_ada # best adaboost model  

AdaBoostClassifier(algorithm='SAMME.R', base_estimator=None, learning_rate=1,
                   n_estimators=300, random_state=None)

In [ ]:
pred_ada = best_ada.predict(X_test)


In [ ]:
accuracy_score(y_test,pred_ada)

0.8185733333333334

here insert f1 score

In [ ]:
from sklearn.metrics import f1_score
f1_score(y_test, pred_ada , average='weighted')


0.8169849852532735

The f1_score acheived for Adaboost is 81.69%

##81% accuracy slighly less compared to Random forest which gave 83% accuracy

##Using SMOTE to upsample the data

we have used smote in notebook 2 to upsample data . Then we have pickled the smote data and ssaved it to drive .In the following cell we are downloading the pickled smote data in this notebook. 

In [ ]:
filename = '/content/drive/MyDrive/Higgs_boson/list_dataset_SMOTE.pickle'
dataset_list = pickle.load(open(filename, 'rb'))
X_train_smote =dataset_list[0]
y_train_smote  =dataset_list[1]

In [ ]:
X_train_smote.shape, y_train_smote.shape # checking if its the correct shape

((230066, 23), (230066, 1))

###Same parameters which we used for the imbalanced data

In [ ]:
boost = AdaBoostClassifier() #adaboost object 

In [ ]:
param_grid = {'n_estimators':[100,200,300], 'learning_rate':[0.01, 0.1, 1],} # dictionary of hyperparameters

In [ ]:
grid = GridSearchCV(boost, param_grid,cv=3,n_jobs=-1 ) # grid object

####Running gridcv fit code on another powerful machine and pickling the grid results then importing them into this google colab Notebook

In [ ]:
start_time = time.time()
grid.fit(X_train_smote,y_train_smote)
stop_time = time.time()
print("Code started at : ",time.ctime(start_time))
print("Code started at : ",time.ctime(stop_time))
print("Total time taken : ",((stop_time-start_time)/60),"minutes")

Above cell ran in another machine for faster computation

In [ ]:
best_ada_smote  = grid.best_estimator_

Since the grid.fit ran in another computer we have pickled results to import them in this notebook

In [ ]:
filename = '/content/drive/MyDrive/Higgs_boson/best_ada_smote.pickle'
best_ada_smote = pickle.load(open(filename, 'rb'))

/usr/local/lib/python3.7/dist-packages/sklearn/base.py:318: UserWarning: Trying to unpickle estimator DecisionTreeClassifier from version 0.23.2 when using version 0.22.2.post1. This might lead to breaking code or invalid results. Use at your own risk.
  UserWarning)
/usr/local/lib/python3.7/dist-packages/sklearn/base.py:318: UserWarning: Trying to unpickle estimator AdaBoostClassifier from version 0.23.2 when using version 0.22.2.post1. This might lead to breaking code or invalid results. Use at your own risk.
  UserWarning)


In [ ]:
best_ada_smote

AdaBoostClassifier(algorithm='SAMME.R', base_estimator=None, learning_rate=1,
                   n_estimators=300, random_state=None)

In [ ]:
pred_ada_smote = best_ada_smote.predict(X_test)

In [ ]:
accuracy_score(y_test,pred_ada_smote)

0.7982666666666667

Accuracy of 79.83% after SMOTE

f1 score

In [ ]:
from sklearn.metrics import f1_score
f1_score(y_test, pred_ada_smote , average='weighted')


0.8014969211526923

The f1_score acheived for Adaboost after SMOTE is 80.15%

##Conclusion

The f1_score acheived for Adaboost is 81.69%

The f1_score acheived for Adaboost after SMOTE is 80.15%

The f1_score acheived for Random Forest after SMOTE is 82.09.

####The F1 score for random forest is 83.0%

Our random forest Model has performed better than Adaboost and with more hyper parameter tunning the accuracy and F1 score can be increased. We think random forest perfect better because its not very sensitive towards imbalanced data 

### We would want to try Xgboost with different base models and neural network for this data set for further improvment 